File: FirebaseDatabase/Sources/Login/FIRDatabaseConnectionContextProvider.m
/*
 * Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FirebaseDatabase/Sources/Login/FIRDatabaseConnectionContextProvider.h"

#import "FirebaseCore/Sources/Private/FirebaseCoreInternal.h"

#import "FirebaseAppCheck/Sources/Interop/FIRAppCheckInterop.h"
#import "FirebaseAppCheck/Sources/Interop/FIRAppCheckTokenResultInterop.h"
#import "Interop/Auth/Public/FIRAuthInterop.h"

#import "FirebaseDatabase/Sources/Api/Private/FIRDatabaseQuery_Private.h"
#import "FirebaseDatabase/Sources/Utilities/FUtilities.h"

NS_ASSUME_NONNULL_BEGIN

@interface FAuthStateListenerWrapper : NSObject

@property(nonatomic, copy) fbt_void_nsstring listener;
@property(nonatomic, weak) id<FIRAuthInterop> auth;

@end

@implementation FAuthStateListenerWrapper

- (instancetype)initWithListener:(fbt_void_nsstring)listener
                            auth:(id<FIRAuthInterop>)auth {
    self = [super init];
    if (self != nil) {
        self->_listener = listener;
        self->_auth = auth;
        [[NSNotificationCenter defaultCenter]
            addObserver:self
               selector:@selector(authStateDidChangeNotification:)
                   name:FIRAuthStateDidChangeInternalNotification
                 object:nil];
    }
    return self;
}

- (void)authStateDidChangeNotification:(NSNotification *)notification {
    NSDictionary *userInfo = notification.userInfo;
    if (notification.object == self.auth) {
        NSString *token =
            userInfo[FIRAuthStateDidChangeInternalNotificationTokenKey];
        dispatch_async([FIRDatabaseQuery sharedQueue], ^{
          self.listener(token);
        });
    }
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

@end

@implementation FIRDatabaseConnectionContext

- (instancetype)initWithAuthToken:(nullable NSString *)authToken
                    appCheckToken:(nullable NSString *)appCheckToken {
    self = [super init];
    if (self) {
        _authToken = [authToken copy];
        _appCheckToken = [appCheckToken copy];
    }
    return self;
}

@end

@interface FIRDatabaseConnectionContextProvider ()

@property(nonatomic, strong) id<FIRAppCheckInterop> appCheck;
@property(nonatomic, strong) id<FIRAuthInterop> auth;

/// Strong references to the auth listeners as they are only weak in
/// FIRFirebaseApp.
@property(nonatomic, readonly) NSMutableArray *authListeners;

/// Observer objects returned by
/// `-[NSNotificationCenter addObserverForName:object:queue:usingBlock:]`
/// method. Required to cleanup the observers on dealloc.
@property(nonatomic, readonly) NSMutableArray *appCheckNotificationObservers;

/// An NSOperationQueue to call listeners on.
@property(nonatomic, readonly) NSOperationQueue *listenerQueue;

@end

@implementation FIRDatabaseConnectionContextProvider

- (instancetype)initWithAuth:(nullable id<FIRAuthInterop>)auth
                    appCheck:(nullable id<FIRAppCheckInterop>)appCheck {
    self = [super init];
    if (self != nil) {
        self->_appCheck = appCheck;
        self->_auth = auth;
        self->_authListeners = [NSMutableArray array];
        self->_appCheckNotificationObservers = [NSMutableArray array];
        self->_listenerQueue = [[NSOperationQueue alloc] init];
        self->_listenerQueue.underlyingQueue = [FIRDatabaseQuery sharedQueue];
    }
    return self;
}

- (void)dealloc {
    @synchronized(self) {
        // Make sure notification observers are removed from
        // NSNotificationCenter.
        for (id notificationObserver in self.appCheckNotificationObservers) {
            [NSNotificationCenter.defaultCenter
                removeObserver:notificationObserver];
        }
    }
}

- (void)
    fetchContextForcingRefresh:(BOOL)forceRefresh
                  withCallback:
                      (void (^)(FIRDatabaseConnectionContext *_Nullable context,
                                NSError *_Nullable error))callback {

    if (self.auth == nil && self.appCheck == nil) {
        // Nothing to fetch. Finish straight away.
        callback(nil, nil);
        return;
    }

    // Use dispatch group to call the callback when both Auth and FAC operations
    // finished.
    dispatch_group_t dispatchGroup = dispatch_group_create();

    __block NSString *authToken;
    __block NSString *appCheckToken;
    __block NSError *authError;

    if (self.auth) {
        dispatch_group_enter(dispatchGroup);
        [self.auth getTokenForcingRefresh:forceRefresh
                             withCallback:^(NSString *_Nullable token,
                                            NSError *_Nullable error) {
                               authToken = token;
                               authError = error;

                               dispatch_group_leave(dispatchGroup);
                             }];
    }

    if (self.appCheck) {
        dispatch_group_enter(dispatchGroup);
        [self.appCheck
            getTokenForcingRefresh:forceRefresh
                        completion:^(
                            id<FIRAppCheckTokenResultInterop> _Nonnull tokenResult) {
                          appCheckToken = tokenResult.token;
                          if (tokenResult.error) {
                              FFLog(@"I-RDB096001",
                                    @"Failed to fetch App Check token: %@",
                                    tokenResult.error);
                          }
                          dispatch_group_leave(dispatchGroup);
                        }];
    }

    dispatch_group_notify(dispatchGroup, [FIRDatabaseQuery sharedQueue], ^{
      __auto_type context = [[FIRDatabaseConnectionContext alloc]
          initWithAuthToken:authToken
              appCheckToken:appCheckToken];
      // Pass only a possible Auth error. App Check errors should not change the
      // database SDK behaviour at this point as the App Check enforcement is
      // controlled on the backend.
      callback(context, authError);
    });
}

- (void)listenForAuthTokenChanges:(_Nonnull fbt_void_nsstring)listener {
    FAuthStateListenerWrapper *wrapper =
        [[FAuthStateListenerWrapper alloc] initWithListener:listener
                                                       auth:self.auth];
    [self.authListeners addObject:wrapper];
}

- (void)listenForAppCheckTokenChanges:(fbt_void_nsstring)listener {
    if (self.appCheck == nil) {
        return;
    }
    NSString *appCheckTokenKey = [self.appCheck notificationTokenKey];
    __auto_type notificationObserver = [NSNotificationCenter.defaultCenter
        addObserverForName:[self.appCheck tokenDidChangeNotificationName]
                    object:self.appCheck
                     queue:self.listenerQueue
                usingBlock:^(NSNotification *_Nonnull notification) {
                  NSString *appCheckToken =
                      notification.userInfo[appCheckTokenKey];
                  listener(appCheckToken);
                }];

    @synchronized(self) {
        [self.appCheckNotificationObservers addObject:notificationObserver];
    }
}

+ (id<FIRDatabaseConnectionContextProvider>)
    contextProviderWithAuth:(nullable id<FIRAuthInterop>)auth
                   appCheck:(nullable id<FIRAppCheckInterop>)appCheck {
    return [[self alloc] initWithAuth:auth appCheck:appCheck];
}

@end

NS_ASSUME_NONNULL_END

File: FirebaseDatabaseSwiftCore/Sources/Login/FIRDatabaseConnectionContextProvider.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 12/04/2022.
//

import Foundation


@objc(FIRDatabaseConnectionContext) public class DatabaseConnectionContext: NSObject {
    /// Auth token if available.
    @objc public var authToken: String?

    /// App check token if available.
    @objc public var appCheckToken: String?

    @objc public init(authToken: String?, appCheckToken: String?) {
        self.authToken = authToken
        self.appCheckToken = appCheckToken
    }
}

@objc(FIRDatabaseConnectionContextProvider) public protocol DatabaseConnectionContextProviderProtocol: NSObjectProtocol {
    func fetchContextForcingRefresh(_ forceRefresh: Bool, withCallback callback: @escaping (DatabaseConnectionContext?, Error?) -> Void)

    /// Adds a listener to the Auth token updates.
    /// @param listener A block that will be invoked each time the Auth token is
    /// updated.
    func listenForAuthTokenChanges(_ listener:  @escaping (String) -> Void)

    /// Adds a listener to the FAC token updates.
    /// @param listener A block that will be invoked each time the FAC token is
    /// updated.
    func listenForAppCheckTokenChanges(_ listener: @escaping (String) -> Void)
}

extension Notification.Name {
    public static let FIRAuthStateDidChangeInternalNotification = Notification.Name("FIRAuthStateDidChangeInternalNotification")
}

let FIRAuthStateDidChangeInternalNotificationTokenKey = "FIRAuthStateDidChangeInternalNotificationTokenKey"

private class FAuthStateListenerWrapper {
    private let listener: (String) -> Void
    private weak var auth: DatabaseAuthInterop?
    private let queue: DispatchQueue

    init(listener: @escaping (String) -> Void, auth: DatabaseAuthInterop, queue: DispatchQueue) {
        self.listener = listener
        self.auth = auth
        self.queue = queue
        NotificationCenter
            .default
            .addObserver(self,
                         selector: #selector(authStateDidChangeNotification),
                         name: .FIRAuthStateDidChangeInternalNotification,
                         object: nil)
    }
    @objc func authStateDidChangeNotification(_ notification: Notification) {
        let userInfo = notification.userInfo
        guard (notification.object as? AnyObject) === self.auth else { return }
        guard let token = userInfo?[FIRAuthStateDidChangeInternalNotificationTokenKey] as? String else { return }
        queue.async {
            self.listener(token)
        }
    }
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}

@objc(FIRDatabaseAuthInterop) public protocol DatabaseAuthInterop: NSObjectProtocol {
    func getTokenForcingRefresh(_ forceRefresh: Bool, withCallback callback: (String?, Error?) -> Void)
}

@objc(FIRDatabaseAppCheckTokenResultInterop) public protocol DatabaseAppCheckTokenResultInterop: NSObjectProtocol {
    var token: String? { get }
    var error: Error? { get }
}

@objc(FIRDatabaseAppCheckInterop) public protocol DatabaseAppCheckInterop: NSObjectProtocol {
    func getTokenForcingRefresh(_ forceRefresh: Bool, completion: @escaping (DatabaseAppCheckTokenResultInterop) -> Void)
    var notificationTokenKey: String { get }
    var tokenDidChangeNotificationName: Notification.Name { get }
}

// TODO: Make FIRAppCheckInterop conform to FIRDatabaseAppCheckInterop
// TODO: Make FIRAppCheckTokenResultInterop conform to FIRDatabaseAppCheckTokenResultInterop
// TODO: Make FIRAuthInterop conform to FIRDatabaseAuthInterop

@objc(FIRDatabaseConnectionContextProvider) public class DatabaseConnectionContextProvider: NSObject, DatabaseConnectionContextProviderProtocol {

    var appCheck: DatabaseAppCheckInterop? // FIRAppCheckInterop
    var auth: DatabaseAuthInterop? // FIRAuthInterop

    /// Strong references to the auth listeners as they are only weak in
    /// FIRFirebaseApp.
    private var authListeners: [FAuthStateListenerWrapper] = []

    /// Observer objects returned by
    /// `-[NSNotificationCenter addObserverForName:object:queue:usingBlock:]`
    /// method. Required to cleanup the observers on dealloc.
    var appCheckNotificationObservers: [Any] = []

    /// An NSOperationQueue to call listeners on.
    var listenerQueue: OperationQueue

    private let dispatchQueue: DispatchQueue

    private init(auth: DatabaseAuthInterop?,
         appCheck: DatabaseAppCheckInterop?,
                 dispatchQueue: DispatchQueue) {
        self.appCheck = appCheck
        self.auth = auth
        self.dispatchQueue = dispatchQueue
        self.listenerQueue = OperationQueue()
        self.listenerQueue.underlyingQueue = dispatchQueue
    }

    deinit {
        // TODO: Will this work on Linux?
        // NOTE: Maybe it doesn't need to. Auth will be likely be bridged
        // in some other way
        // Otherwise we need some other synchronization method
        objc_sync_enter(self)
        defer { objc_sync_exit(self) }
        for observer in self.appCheckNotificationObservers {
            NotificationCenter.default.removeObserver(observer)
        }
    }

    public func fetchContextForcingRefresh(_ forceRefresh: Bool, withCallback callback: @escaping (DatabaseConnectionContext?, Error?) -> Void) {
        guard self.auth != nil || self.appCheck != nil else {
            // Nothing to fetch. Finish straight away.
            // XXX TODO: HACK TO MAKE TESTING WORK
            callback(DatabaseConnectionContext(authToken: nil, appCheckToken: nil), nil)
//            callback(nil, nil)
            return
        }
        // Use dispatch group to call the callback when both Auth and FAC operations
        // finished.
        let dispatchGroup = DispatchGroup()
        var authToken: String? = nil
        var appCheckToken: String? = nil
        var authError: Error? = nil
        if let auth = auth {
            dispatchGroup.enter()
            auth.getTokenForcingRefresh(forceRefresh) { token, error in
                authToken = token
                authError = error
                dispatchGroup.leave()
            }
        }
        if let appCheck = appCheck {
            dispatchGroup.enter()
            appCheck.getTokenForcingRefresh(forceRefresh) { tokenResult in
                appCheckToken = tokenResult.token
                if let error = tokenResult.error {
                    FFLog("I-RDB096001", "Failed to fetch App Check token: \(error)")
                }
                dispatchGroup.leave()
            }
        }
        dispatchGroup.notify(queue: dispatchQueue, execute: {
            let context = DatabaseConnectionContext(authToken: authToken, appCheckToken: appCheckToken)
            // Pass only a possible Auth error. App Check errors should not change the
            // database SDK behaviour at this point as the App Check enforcement is
            // controlled on the backend.
            callback(context, authError)
        })
    }

    public func listenForAuthTokenChanges(_ listener: @escaping (String) -> Void) {
        guard let auth = auth else {
            return
        }

        let wrapper = FAuthStateListenerWrapper(listener: listener, auth: auth, queue: dispatchQueue)
        authListeners.append(wrapper)
    }

    public func listenForAppCheckTokenChanges(_ listener: @escaping (String) -> Void) {
        guard let appCheck = appCheck else {
            return
        }
        let appCheckTokenKey = appCheck.notificationTokenKey
        let observer = NotificationCenter.default
            .addObserver(forName: appCheck.tokenDidChangeNotificationName,
                         object: appCheck,
                         queue: listenerQueue) { notification in
                guard let appCheckToken = notification.userInfo?[appCheckTokenKey] as? String else {
                    return
                }
                listener(appCheckToken)
            }

        objc_sync_enter(self)
        defer { objc_sync_exit(self) }
        self.appCheckNotificationObservers.append(observer)
    }

    @objc public class func contextProvider(auth: DatabaseAuthInterop?, appCheck: DatabaseAppCheckInterop?, dispatchQueue: DispatchQueue) -> DatabaseConnectionContextProviderProtocol {
        DatabaseConnectionContextProvider(auth: auth, appCheck: appCheck, dispatchQueue: dispatchQueue)
    }
}
